File: host.py
import tornado.httpserver
import tornado.websocket
import tornado.ioloop
from tornado.concurrent import return_future
from tornado.gen import coroutine
from tornado.httputil import HTTPHeaders
import tornado.web
import socket
import json

import os
import sys
from bs4 import BeautifulSoup
import requests

def extract_car_data(data):
    try:
        print data

        for car in data.get("results",[]):
            plate = {
                "confidence": car.get("confidence"),
                "name": car.get("plate")
            }
            colour = car.get("vehicle",{}).get("color")[0] #{confidence, name}
            make = car.get("vehicle",{}).get("make")[0] #{confidence, name}
            body_type = car.get("vehicle",{}).get("body_type")[0] #{confidence, name}
            year = car.get("vehicle",{}).get("year")[0] #{confidence, name} (range)
            model = {
                "confidence": car.get("vehicle",{}).get("make_model")[0].get("confidence"),
                "name": "_".join(car.get("vehicle",{}).get("make_model")[0].get("name","").split("_")[1:])
            }

            obj = {
                "plate": plate,
                "colour": colour,
                "year": year,
                "body_type": body_type,
                "make": make,
                "model": model
            }
            print obj
            return obj
    except Exception as ex:
        print "{}".format(ex)
        return {}
    else:
        return {}

def extract_plate_data(reg):
    try:
        if not reg:
            return {}

        #create session
        s = requests.Session()

        #login
        url = "https://www.ecom.transport.sa.gov.au/et/checkRegistrationExpiryDate.do"
        r = s.get(url)

        #query
        url = "https://www.ecom.transport.sa.gov.au/et/cred_registration_details.do"
        data = {"plateNumber": reg.upper()}
        r = s.post(url, data)
        soup = BeautifulSoup(r.text, 'html.parser')

        #get data
        def clean(str):
            return str.encode('utf-8').replace('\xc2\xa0', '_').lower().strip()
        domData = dict(zip(soup.find_all('td', {'class':'fieldTitle'}),soup.find_all('td', {'class':'fieldSpace'})))
        data = {clean(k.text):clean(v.text) for (k,v) in domData.iteritems()}

        if "please_enter_a_south_australian_plate_number" in data:
            return None

        #use data
        return data
    except Exception as ex:
        print "{}".format(ex)
        return {}

class RESTHandler(tornado.web.RequestHandler):
    def initialize(self, **kwargs):
        pass

    def options(self, *args):
        allowedMethods = list(["OPTIONS", "GET"])
        self.set_header("Access-Control-Allow-Methods", ",".join(allowedMethods))
        self.finish()

    @coroutine
    def post(self):

        #get post data
        input_json = tornado.escape.json_decode(self.request.body)
        img_base64 = input_json.get("img_src")
        _, img_base64_data = img_base64.split(',')

        #make request
        SECRET_KEY = os.environ.get('SECRET_KEY','')
        url = 'https://api.openalpr.com/v2/recognize_bytes?recognize_vehicle=1&country=au&secret_key=%s' % (SECRET_KEY)
        r = requests.post(url, data = img_base64_data)

        #extract car data
        print "Response from recognition lookup: {}".format(r.status_code)
        car_data = extract_car_data(r.json() or {})

        #fetch/extract plate data from car_data.plate.name
        plate_data = extract_plate_data(car_data.get("plate",{}).get("name") if car_data else None)

        data = {
            "car": car_data,
            "plate": plate_data
        }

        if any(data):
            self.set_status(200)
            self.finish(data)
        else:
            self.set_status(400)
            self.finish()

def main():
    application = tornado.web.Application([
        (r'/scan', RESTHandler),
        (r"/(.*)", tornado.web.StaticFileHandler, {"path": 'static', "default_filename": "index.html"})
    ])
    http_server = tornado.httpserver.HTTPServer(application)
    port = int(os.environ.get("PORT", 8080))
    http_server.listen(port)
    tornado.ioloop.IOLoop.current().start()

if __name__ == "__main__":
    main()

File: host.js
const express = require('express')
const multer  = require('multer')
const app = express()
const path = require('path')

const request = require('request')
const imagemin = require('imagemin')
const imageminPngquant = require('imagemin-pngquant')
var upload = multer()

var plateToPlateInformation = require('./ezyreg.js')

// app.get('/', (req, res) => res.send('Hello World!'))
app.use('/', express.static(path.join(__dirname, 'static')))

var compressImageBuffer = function(buffer, quality) {
    console.log("Compressing Image Buffer")
    console.log("Buffer size "+buffer.length/1024+"kb")
    return new Promise(function(resolve, reject) {
        imagemin.buffer(buffer, {plugins: [imageminPngquant({quality:quality})]})
            .then(function(buffer) {
                console.log("Compressing Image Buffer, resolve()")
                resolve(buffer)
            })
            .catch(function(error){
                console.log("Compressing Image Buffer, reject()")
                reject(error)
            })
    })
}

var imageBufferToBase64 = function(buffer) {
    console.log("Converting Image Buffer to Base64")
    console.log("Buffer size "+buffer.length/1024+"kb")
    return new Promise(function(resolve, reject) {
        try {
            var base64 = buffer.toString("base64")
            console.log("Converting Image Buffer to Base64, resolve()")
            resolve(base64);
        } catch(ex) {
            console.log("Converting Image Buffer to Base64, reject()")
            reject(null);
        }
    })
}

var base64ToPlate = function(base64) {
    console.log("Converting Base64 Image to Plate Recognition")
    console.log(base64)
    return new Promise(function(resolve, reject) {
        var secret_key = process.env.SECRET_KEY || '';
        var url = 'https://api.openalpr.com/v2/recognize_bytes?limit=1&recognize_vehicle=0&country=au&secret_key=';
        request.post({'url':url + secret_key, body: base64},
            function(error, response, body) {
                try {
                    let json = JSON.parse(body)
                    console.log("Converting Base64 Image to Plate Recognition, resolve()")
                    console.log("Response: "+body)
                    resolve({
                        number: json["results"][0]["plate"],
                        confidence: json["results"][0]["confidence"]
                    })
                } catch (ex) {
                    console.log("Converting Base64 Image to Plate Recognition, resolve(null)")
                    resolve(null)
                }
            }
        )
    })
}

app.post('/scan', upload.single('file'), function (req, res, next) {
    compressImageBuffer(req.file.buffer, '50')
        .then(imageBufferToBase64)
        .then(base64ToPlate)
        .then(plateToPlateInformation)
        .then(function(plateInformation) {
            if (plateInformation) {
                res.json(plateInformation)
            } else {
                res.status(404).send()
            }
        })
        .catch(function(error) {
            console.log("/scan 500")
            console.error(error)
            res.status(500).send()
        })
})

app.get('/reg/:plate', function(req, res) {
    plateToPlateInformation({"number":req.params.plate})
        .then(function(plateInformation) {
            res.json(plateInformation)
        })
        .catch(function() {
            res.status(500).send()
        })
});

var p = process.env.PORT || 8080;
app.listen(p, () => console.log('Example app listening on port '+p))